### DISCUSSION TITLE
Using Inquirer with piped input

### DISCUSSION THREAD

**Discussion by bschick** [SBoudrias/Inquirer.js#1605] [Q&A]
I'd like to create a tool that takes input from a pipe and then prompts the user for more input when needed. I haven't found a way to release stdin from the pipe so the prompts works, however. Tried `process.stdin.unpipe()` but that doesn't do the trick.

Another idea was to close and reopen stdin to connect it to the tty, but its unclear to me if that's possible in nodejs. Example of what doesn't work below. Any suggestions?

> $ echo 'hi' | node prompt.js

```
import { readFileSync } from 'fs';
import { input } from '@inquirer/prompts';

async function main() {
   try {
      console.log(readFileSync(process.stdin.fd, 'utf-8'));
      process.stdin.unpipe();
   } catch (err) { }

   console.log(await input({ message: 'Enter your name' }));
}

await main();
console.log('bye');
```
Output is

>            reject(new ExitPromptError(`User force closed the prompt with ${code} ${signal}`));
>                    ^
>     ExitPromptError: User force closed the prompt with 0 null

**Comment by SBoudrias** [comment]
My guess is piping `echo 'hi'` will send `hi`, followed by an exit signal once the content is exhausted.

Maybe you could filter out the exit signal with a piped stream, and pass that stream to inquirer. Dummy code example, the implementation of the filtering is likely different:

```js
const stdin = process.stdin.pipe(ignoreExitSignal);

await input({ ... }, { stdin });
```

**Comment by bschick** [comment]
There was actually a fairly simple solution, the[ reopen-tty](https://www.npmjs.com/package/reopen-tty) package provides a cross-platform way to open the tty directly. Then its just a matter of passing a custom input stream to prompt rather than stdin which may be attached to a pipe...

```
 reopenTTY.stdin(async (err, ttyStream) => {
   console.log(await input(
      { message: 'Enter your name' },
      { input: ttyStream }
   ));
}
```

It would be nice if there is a way specify a custom input stream for all prompts rather than passing context to each call.

**Reply by SBoudrias** [reply]
> It would be nice if there is a way specify a custom input stream for all prompts rather than passing context to each call.

Yeah, I'm not against it. But we'll need to think about it more deeply. What's the impact of the design we come up with on the community prompts? What's the backward compatibility story? And to which extent authors really expect it to be always global; like is there valid use cases where we want to by-pass this?
